filament/backend/src/vulkan/caching/VulkanDescriptorSetManager.cpp

```diff
@@ -374,632 +246,208 @@ struct Equal {
-            if (texture) {
-                mResources.release(texture);
+        for (uint8_t i = 0; i < UNIQUE_DESCRIPTOR_SET_COUNT; ++i) {
+            if ((setMask & (1LL << i))  == 0) {
+                continue;
+            }
+            auto const vkset = mStashedSets[i];
+            if (auto itr = mHistory.find(vkset); itr != mHistory.end() && !itr->second.bound()) {
+                allBound = false;
+                iterators[i] = itr;
             }
-            mSamplerMap[i] = {{}, nullptr};
         }
-        mInputAttachment = {};
-        mHaveDynamicUbos = false;
-
-        FVK_SYSTRACE_END();
-        return pipelineLayout;
-    }
-
-    void dynamicBind(VulkanCommandBuffer* commands, Handle<VulkanDescriptorSetLayout> uboLayout) {
-        if (!mHaveDynamicUbos) {
+        BoundInfo nextInfo = {
+            pipelineLayout,
+            setMask,
+            mStashedSets,
+        };
+        if (allBound && mLastBoundInfo == nextInfo) {
             return;
         }
-        FVK_SYSTRACE_CONTEXT();
-        FVK_SYSTRACE_START("dynamic-bind");
-
-        assert_invariant(mBoundState.valid());
-        assert_invariant(commands->buffer() == mBoundState.cmdbuf);
-
-        auto layout = mAllocator->handle_cast<VulkanDescriptorSetLayout*>(
-                mBoundState.layouts[UBO_SET_ID]);
-
-        // Note that this is costly, instead just use dynamic UBOs with dynamic offsets.
-        auto const& [set, cached] = getSet(UBO_SET_ID, layout);
-        VkDescriptorSet const vkSet = set->vkSet;
-
-        if (!cached) {
-            VkWriteDescriptorSet descriptorWrites[MAX_UBO_BINDING];
-            uint8_t nwrites = 0;
-
-            for (uint8_t binding: layout->bindings.ubo) {
-                auto const& [info, ubo] = mUboMap[binding];
-                descriptorWrites[nwrites++] = {
-                        .sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET,
-                        .pNext = nullptr,
-                        .dstSet = vkSet,
-                        .dstBinding = binding,
-                        .descriptorCount = 1,
-                        .descriptorType = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
-                        .pBufferInfo = ubo ? &info : &mPlaceHolderBufferInfo,
-                };
-                if (ubo) {
-                    set->resources.acquire(ubo);
-                }
-            }
-            if (nwrites > 0) {
-                vkUpdateDescriptorSets(mDevice, nwrites, descriptorWrites, 0, nullptr);
-            }
-        }
-        commands->acquire(set);
 
-        if (mBoundState.vkSets[UBO_SET_ID] != vkSet) {
-            vkCmdBindDescriptorSets(mBoundState.cmdbuf, VK_PIPELINE_BIND_POINT_GRAPHICS,
-                    mBoundState.pipelineLayout, 0, 1, &vkSet, 0, nullptr);
-            mBoundState.vkSets[UBO_SET_ID] = vkSet;
+        for (uint8_t i = 0; i < UNIQUE_DESCRIPTOR_SET_COUNT; ++i) {
+            auto& itr = iterators[i];
+            if (itr != mHistory.end()) {
```

Comment:
it seems to me that all this code could be replaced by a 4 bits bitset with `forEachBitSet()`. All this does above is figure which of the 4 set needs a call to `bind()`, so this should just be 4 bits.

This code is part of the most inner loop we have, it should be optimized to death!

There should also be a fast path for "just changing the offsets". The way I did this in the gl backend is that I keep two sets of bitset and I can very quickly know which sets (1) don't change, (2) are only a bind offset update or  (3) need a real rebind.